### PR TITLE
Support slices of foreign keys

### PR DESCRIPTION
**Pull request by dgpc**
- [X] Serialize Go slices/maps of refs into `[]string` and `map[string]string`
- [x] Unserialize `[]string` into Go structs on refs
- [x] Implement expand functionality for same

### CODE REVIEW

File: end_to_end_test.go [hunk 1]
```diff
@@ -112,9 +123,16 @@ func TestSetAndGetByID(t *testing.T) {
...
+	var zeroTime time.Time // clear timestamps for comparison
+	for i := range savedEvent.Attendees {
+		savedEvent.Attendees[i].CreateTime = zeroTime
+		savedEvent.Attendees[i].UpdateTime = zeroTime
+	}
+
 	assert.Equal(t, eventRef.ID, savedEvent.ID)
 	assert.NotZero(t, savedEvent.CreateTime)
 	assert.Equal(t, savedEvent.CreateTime, savedEvent.UpdateTime)
+	assert.Equal(t, []User{bilbo, gandalf, thorin}, savedEvent.Attendees)
```
Comment (evnp) [inline end_to_end_test.go:135]: It's a good thing go avoids BDD-style testing, you can't `expect` this party ;)

File: doc_to_model.go [hunk 2]
```diff
@@ -102,7 +102,21 @@ func dataToValue(v reflect.Value, d interface{}) error {
...
 		return populateMap(v, x)
 	case reflect.Slice:
-		return errors.New("calcifer: slice deserialization unimplemented")
+		if dv.Kind() != reflect.Slice {
+			return typeErr()
+		}
+		dlen := dv.Len()
+		vlen := v.Len()
+		if vlen < dlen {
+			v.Set(reflect.MakeSlice(v.Type(), dlen, dlen))
```
Comment (evnp) [inline doc_to_model.go:111]: I'm trying to figure out why `dlen` is repeated here, I gather (from docs) that one is the length and the other is capacity but would the latter be in bytes? The docs on this are minimal.. I'm sure it's obvious to someone more familiar with go

File: doc_to_model.go [hunk 3]
```diff
@@ -198,11 +218,49 @@ func populateMap(v reflect.Value, d map[string]interface{}) error {
 	return errors.New("calcifer: populateMap: unimplemented")
 }
 
-func populateForeignKey(v reflect.Value, d string) error {
+func populateForeignKey(v reflect.Value, dd interface{}) error {
+	d, ok := dd.(string)
```
Comment (evnp) [inline doc_to_model.go:222]: dd -> dereferenced data? (going off earlier dv)

File: doc_to_model.go (hunk 2 above)
Comment (dgpc) [reply doc_to_model.go:111]: this is the reflection equivalent of `make()`, so both length and capacity are in number of elements, not bytes

`reflect.MakeSlice(t, a, b)` is roughly equivalent to `make([]t, a, b)`

space is allocated for `b` `t`s, but `len()` will report `a`.. you can resize/append up to `b` without any memory allocation

in the non-reflect version, the second numeric argument to `make` is optional; and defaults to the first, so the commonly-seen `make([]t, a)` is equivalent to `make([]t, a, a)`,

Comment (evnp) [reply doc_to_model.go:111]: Thanks! makes sense

File: doc_to_model.go (hunk 3 above)
Comment (dgpc) [reply doc_to_model.go:219]: document data 🤷

**Pull request merged**